### DISCUSSION TITLE
How does Bootstrap.DevServer work?

### DISCUSSION THREAD

**Discussion by AaronLS** [unoplatform/Uno.Wasm.Bootstrap#429] [Q&A] [answered]
In the simplest example of using Uno.Wasm.Bootstrap, a console app has Program.cs which just calls Console.Writeline, but somehow a self hosted http listener is initialized.  

There's no code that's initializing any kind of self hosting.  Usually with other self hosted solutions the listener would not start up automatically.  I'd call from Program.cs an API such as WebHostBuilder to initialize the host and begin listening on HTTP.  

I'm trying to understand how does Bootstrap.DevServer hook into the console application startup?  Is all that magic in the *.targets file?

What I'm really trying to work towards is integrating this with a ASP.NET Core MVC project.  Right now just as a proof of concept I am running the MVC self hosted .NET 5 MVC app, and I manually copied the files from the Uno.Bootstrap console app's /bin/Debug/net5.0/dist/ folder into the MVC's wwwrott, and updated my MVC index.cshtml with the references copied from /dist/index.html as below.  This sort of works, but anytime I rebuild the Uno console project, the package name changes.

I'm not sure how to integrate this properly.  I assume my project generating the WASM assemblies needs to remain a separate project from my MVC ASP.NET Core, since I wouldn't want those server assemblies to generate WASM assemblies.  I.e. I don't think I'd reference Uno.Bootstrap from my MVC app.

So it seems like I need some way to make the WASM package names not change everytime it is built, otherwise my references will break every build.  Otherwise I could work out a copy command in a post build step to copy the Uno project's /dist/ output into the MVC project's /wwwroot.

Is there a way to control that package name generation?

Also, there's no special requirement of the server that serves this WASM modules correct? Once compilation is complete at runtime, everything is just being served to the browser as static files correct?

```
@{
    ViewData["Title"] = "Home Page";
}

@section head{

    <script type="text/javascript" src="./package_ac1f45f3abc4ba8c253a4d9a8b35ce86c1089c45/require.js"></script>
    <script type="text/javascript" src="./package_ac1f45f3abc4ba8c253a4d9a8b35ce86c1089c45/mono-config.js"></script>
    <script type="text/javascript" src="./package_ac1f45f3abc4ba8c253a4d9a8b35ce86c1089c45/uno-config.js"></script>
    <script type="text/javascript" src="./package_ac1f45f3abc4ba8c253a4d9a8b35ce86c1089c45/uno-bootstrap.js"></script>
    <script async type="text/javascript" src="./package_ac1f45f3abc4ba8c253a4d9a8b35ce86c1089c45/dotnet.js"></script>
    <link rel="stylesheet" type="text/css" href="./package_ac1f45f3abc4ba8c253a4d9a8b35ce86c1089c45/normalize.css" />
    <link rel="stylesheet" type="text/css" href="./package_ac1f45f3abc4ba8c253a4d9a8b35ce86c1089c45/uno-bootstrap.css" />

}
...
```

**Comment by AaronLS** [comment, marked as answer]
I was able to use the WasmShellOutputDistPath to hard code a push into the other project's wwwroot, and then scan the wwwroot for the newest directory for the references.

In proj for the WASM project:
```

  <Target Name="AfterBuildValidation" AfterTargets="Build">
    <Message Importance="high" Text="After Build Start" />
    <Message Importance="high" Text="WASM Out Path: $(WasmShellOutputPackagePath)" />
   
    <ItemGroup>
        <WasmDistFiles Include="$(WasmShellOutputDistPath)\**\*.*"/>
    </ItemGroup>
    <Message Importance="high" Text="WasmDistFiles: @(WasmDistFiles)" />
    <Copy
          SourceFiles="@(WasmDistFiles)"
          DestinationFiles="@(WasmDistFiles->'$(SolutionDir)\WasmPrototype.Mvc\wwwroot\%(RecursiveDir)%(Filename)%(Extension)')"
          SkipUnchangedFiles="true" />    
		<Message Importance="high" Text="After Build End" />
 
```

And then in my cshtml:

```
{
    var directories = new System.IO.DirectoryInfo(HostingEnvironment.WebRootPath).GetDirectories("package_*").OrderByDescending(d => d.CreationTimeUtc);    
    string wasmPackageName = directories.First().Name;
}

@section head{
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent" />
    <script type="text/javascript" src="./@(wasmPackageName)/require.js"></script>
    <script type="text/javascript" src="./@(wasmPackageName)/mono-config.js"></script>
    <script type="text/javascript" src="./@(wasmPackageName)/uno-config.js"></script>
    <script type="text/javascript" src="./@(wasmPackageName)/uno-bootstrap.js"></script>
    <script async type="text/javascript" src="./@(wasmPackageName)/dotnet.js"></script>
    <link rel="stylesheet" type="text/css" href="./@(wasmPackageName)/normalize.css" />
    <link rel="stylesheet" type="text/css" href="./@(wasmPackageName)/uno-bootstrap.css" />
}
```

**Reply by jeromelaban** [reply]
The `WasmShellOutputDistPath` is indeed the right way to change this, added to the use of the [`WasmShellWebAppBasePath` ](https://github.com/unoplatform/Uno.Wasm.Bootstrap#support-for-deep-linking--routes) property to change the root of the app.

**Reply by AaronLS** [reply]
@jeromelaban Ah I see, thank you.
